src/slash-commands/join.ts

```diff
@@ -0,0 +1,163 @@
+    if (params.length < 1) return;
+
+    let isPermalink = false;
+    if (params[0].startsWith("http:") || params[0].startsWith("https:")) {
+        // It's at least a URL - try and pull out a hostname to check against the
+        // permalink handler
+        const parsedUrl = new URL(params[0]);
+        const hostname = parsedUrl.host || parsedUrl.hostname; // takes first non-falsey value
+
+        // if we're using a Element permalink handler, this will catch it before we get much further.
+        // see below where we make assumptions about parsing the URL.
+        if (isPermalinkHost(hostname)) {
+            isPermalink = true;
+        }
+    }
+
+    if (params[0][0] === "#") {
+        let roomAlias = params[0];
+        if (!roomAlias.includes(":")) {
+            roomAlias += ":" + cli.getDomain();
+        }
+
+        dis.dispatch<ViewRoomPayload>({
+            action: Action.ViewRoom,
+            room_alias: roomAlias,
+            auto_join: autoJoin,
+            metricsTrigger: "SlashCommand",
+            metricsViaKeyboard: true,
+        });
+        return success();
+    }
+
+    if (params[0][0] === "!") {
+        const [roomId, ...viaServers] = params;
+
+        dis.dispatch<ViewRoomPayload>({
+            action: Action.ViewRoom,
+            room_id: roomId,
+            via_servers: viaServers, // for the rejoin button
+            auto_join: autoJoin,
+            metricsTrigger: "SlashCommand",
+            metricsViaKeyboard: true,
+        });
+        return success();
+    }
+
+    if (isPermalink) {
+        const permalinkParts = parsePermalink(params[0]);
+
+        // This check technically isn't needed because we already did our
+        // safety checks up above. However, for good measure, let's be sure.
+        if (!permalinkParts) {
+            return;
+        }
+
+        // If for some reason someone wanted to join a user, we should
+        // stop them now.
+        if (!permalinkParts.roomIdOrAlias) {
+            return;
+        }
+
+        const entity = permalinkParts.roomIdOrAlias;
+        const viaServers = permalinkParts.viaServers;
+        const eventId = permalinkParts.eventId;
+
+        const dispatch: ViewRoomPayload = {
+            action: Action.ViewRoom,
+            auto_join: autoJoin,
+            metricsTrigger: "SlashCommand",
+            metricsViaKeyboard: true,
+        };
+
+        if (entity[0] === "!") dispatch["room_id"] = entity;
+        else dispatch["room_alias"] = entity;
+
+        if (eventId) {
+            dispatch["event_id"] = eventId;
+            dispatch["highlighted"] = true;
+        }
+
+        if (viaServers) {
+            // For the join, these are passed down to the js-sdk's /join call
+            dispatch["opts"] = { viaServers };
+
+            // For if the join fails (rejoin button)
+            dispatch["via_servers"] = viaServers;
+        }
+
+        dis.dispatch(dispatch);
+        return success();
+    }
+}
```

Comment:
it'd be nice to make it explicit that we are deliberately returning undefined here.

```suggestion

    // Otherwise, it's a usage error. Return `undefined`.
}
```